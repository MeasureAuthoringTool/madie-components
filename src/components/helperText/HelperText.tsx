import React from "react";
import styled from "styled-components";
import tw from "twin.macro";

export interface HelperTextProps {
  text: string;
}

export function HelperTextComponent(props: HelperTextProps) {
  const { text, ...args } = props;
  return (
    <p {...args} tw="mt-2 text-sm text-gray-500">
      {text}
    </p>
  );
}
