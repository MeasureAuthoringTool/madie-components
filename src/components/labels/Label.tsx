import React from "react";
import tw, { styled } from "twin.macro";

export interface LabelProps
  extends React.LabelHTMLAttributes<HTMLLabelElement> {
  text: string;
  isOptional?: boolean;
}

const StyledLabel = styled.label(({}) => [
  tw`block text-sm font-medium text-gray-700`,
]);

const StyledSpan = styled.span(({}) => [tw`text-sm text-gray-500`]);

export function LabelComponent(props: LabelProps) {
  const { isOptional, ...args } = props;
  return (
    <div tw="flex justify-between">
      <StyledLabel {...args}>{props.text}</StyledLabel>
      {!!isOptional && <StyledSpan>Optional</StyledSpan>}
    </div>
  );
}
