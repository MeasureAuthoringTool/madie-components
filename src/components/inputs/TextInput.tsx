import React from "react";
import tw, { styled } from "twin.macro";
import { LabelComponent } from "../labels/Label";
import { HelperTextComponent } from "../helperText/HelperText";
import { ExclamationCircleIcon } from "@heroicons/react/solid";
import { MailIcon } from "@heroicons/react/solid";
import { QuestionMarkCircleIcon } from "@heroicons/react/solid";

export interface TextInputProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  children?: any;
  hasError?: boolean;
  isEmail?: boolean;
  hasHelperIcon?: boolean;
}

interface HandleTextInputProps {
  hasError?: boolean;
}

const StyledTextInput = styled.input<HandleTextInputProps>(({ hasError }) => [
  hasError
    ? tw`block w-full pr-10 border-red-300 text-red-900 placeholder-red-300 focus:outline-none focus:ring-red-500 focus:border-red-500 sm:text-sm rounded-md`
    : tw`shadow-sm focus:ring-red-900 focus:border-red-900 block w-full sm:text-sm border-gray-300 rounded-md border-2`,
]);

export function TextInputComponent(props: TextInputProps) {
  const { children, hasError, isEmail, hasHelperIcon, ...args } = props;
  return (
    <div>
      {React.Children.map(children, (child) => {
        if (child.type === LabelComponent) return child;
      })}
      <div tw="mt-1">
        {/* {!!isEmail ? (
          <div tw="absolute inset-y-0 left-0 pl-3 flex items-center pointer-events-none">
            <MailIcon tw="h-5 w-5 text-gray-400" aria-hidden="true" />
          </div>
        ) : (
          ""
        )} */}

        <StyledTextInput {...args} hasError={hasError} />

        {/* {hasHelperIcon ? (
          <div className="absolute inset-y-0 right-0 pr-3 flex items-center pointer-events-none">
            <QuestionMarkCircleIcon
              className="h-5 w-5 text-gray-400"
              aria-hidden="true"
            />
          </div>
        ) : (
          ""
        )} */}
        {/* {!!hasError ? (
          <div tw="absolute inset-y-0 right-0 pr-3 flex items-center pointer-events-none">
            <ExclamationCircleIcon
              tw="h-5 w-5 text-red-500"
              aria-hidden="true"
            />
          </div>
        ) : (
          ""
        )} */}
      </div>
      {React.Children.map(children, (child) => {
        if (child.type === HelperTextComponent) return child;
      })}
    </div>
  );
}
