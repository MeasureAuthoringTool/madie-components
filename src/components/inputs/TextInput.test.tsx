import * as React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { TextInputComponent } from "./TextInput";
import { LabelComponent } from "../labels/Label";
import { HelperTextComponent } from "../helperText/HelperText";
import { ExclamationCircleIcon, MailIcon } from "@heroicons/react/solid";

const textInputTestId = "text-input-component";
const labelProps = {
  testId: "label-component-test",
  text: "Label",
  cornerHint: "Optional",
};

const textInputProps: any = {
  type: "text",
  name: "TextInputName",
  id: "TextInputId",
  placeholder: "you@example.com",
};

const helperTextProps = {
  testId: "helper-text-component-test",
  text: "Help text goes here",
  isError: false,
};

describe("TextInput Component", () => {
  it("should render default text input with no children", () => {
    render(
      <TextInputComponent data-testid={textInputTestId} {...textInputProps} />
    );
    const textInput = screen.getByTestId(textInputTestId);
    expect(textInput).toBeInTheDocument();
    expect(textInput).toHaveAttribute("type", "text");

    userEvent.type(textInput, "test@mail.com");
    expect(textInput).toHaveValue("test@mail.com");

    expect(screen.queryByTestId(labelProps.testId)).not.toBeInTheDocument();
    expect(textInput).toMatchSnapshot();
  });

  it("should render text input with children components", () => {
    render(
      <TextInputComponent data-testid={textInputTestId} {...textInputProps}>
        <LabelComponent
          text={labelProps.text}
          cornerHint={labelProps.cornerHint}
        />
        <HelperTextComponent
          data-testid={helperTextProps.testId}
          text={helperTextProps.text}
        />
      </TextInputComponent>
    );
    const textInput = screen.getByTestId(textInputTestId);
    expect(textInput).toBeInTheDocument();
    expect(textInput).toHaveAttribute("type", "text");
    expect(textInput).toMatchSnapshot();

    const label = screen.queryByTestId(labelProps.testId);
    expect(label).toBeInTheDocument();
    expect(label).toHaveTextContent(labelProps.text);
    expect(label).toHaveTextContent(labelProps.cornerHint);
    expect(label).toMatchSnapshot();

    const helperText = screen.queryByTestId(helperTextProps.testId);
    expect(helperText).toBeInTheDocument();
    expect(helperText).toHaveTextContent(helperTextProps.text);
    expect(helperText).toMatchSnapshot();
  });

  textInputProps.leftIcon = <MailIcon data-testid={"left-icon-test"} />;
  textInputProps.rightIcon = (
    <ExclamationCircleIcon data-testid={"right-icon-test"} />
  );
  it("should render text input with left icon and right icon", () => {
    render(<TextInputComponent {...textInputProps} />);
    const leftIcon = screen.getByTestId("left-icon-test");
    expect(leftIcon).toBeInTheDocument();
    expect(leftIcon).toMatchSnapshot();

    const rightIcon = screen.getByTestId("right-icon-test");
    expect(rightIcon).toBeInTheDocument();
    expect(rightIcon).toMatchSnapshot();
  });

  textInputProps.hasError = true;
  it("should render text input with error", () => {
    render(<TextInputComponent {...textInputProps} />);
    const errorIcon = screen.getByTestId("error-test-id");
    expect(errorIcon).toBeInTheDocument();
    expect(errorIcon).toMatchSnapshot();
  });

  textInputProps.isValidationSuccess = true;
  it("should render text input with success validation", () => {
    render(<TextInputComponent {...textInputProps} />);
    const successIcon = screen.getByTestId("success-test-id");
    expect(successIcon).toBeInTheDocument();
    expect(successIcon).toMatchSnapshot();
  });
});
