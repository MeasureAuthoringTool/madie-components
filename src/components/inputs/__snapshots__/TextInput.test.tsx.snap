// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Component should render default text input with no children 1`] = `
<input
  class="TextInput__defaultTextInput-sc-qqk5wt-0 TextInput__styleInput-sc-qqk5wt-1 TextInput__styleInput-sc-qqk5wt-2 TextInput__styleInput-sc-qqk5wt-3 keBmkK dJZKow hLynPY cCFASB"
  data-testid="text-input-component"
  id="TextInputId"
  name="TextInputName"
  placeholder="you@example.com"
  type="text"
/>
`;

exports[`TextInput Component should render text input with children components 1`] = `
<input
  class="TextInput__defaultTextInput-sc-qqk5wt-0 TextInput__styleInput-sc-qqk5wt-1 TextInput__styleInput-sc-qqk5wt-2 TextInput__styleInput-sc-qqk5wt-3 keBmkK dJZKow hLynPY cCFASB"
  data-testid="text-input-component"
  id="TextInputId"
  name="TextInputName"
  placeholder="you@example.com"
  type="text"
/>
`;

exports[`TextInput Component should render text input with children components 2`] = `
<div
  data-testid="label-component-test"
  tw="flex justify-between"
>
  <label
    text="Label"
    tw="block text-sm font-medium text-gray-800 dark:text-gray-300"
  >
    Label
  </label>
  <span
    tw="text-sm text-gray-600 dark:text-gray-300"
  >
    Optional
  </span>
</div>
`;

exports[`TextInput Component should render text input with children components 3`] = `
<span
  class="HelperText__defaultHelperText-sc-4k037i-0 gIFihb"
  data-testid="helper-text-component-test"
>
  Help text goes here
</span>
`;

exports[`TextInput Component should render text input with error 1`] = `
<div
  class="TextInput___StyledDiv6-sc-qqk5wt-9 kLbAnK"
  data-testid="error-test-id"
>
  <svg
    aria-hidden="true"
    class="TextInput___StyledExclamationCircleIcon-sc-qqk5wt-10 jLFXnf"
    fill="currentColor"
    viewBox="0 0 20 20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clip-rule="evenodd"
      d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-7 4a1 1 0 11-2 0 1 1 0 012 0zm-1-9a1 1 0 00-1 1v4a1 1 0 102 0V6a1 1 0 00-1-1z"
      fill-rule="evenodd"
    />
  </svg>
</div>
`;

exports[`TextInput Component should render text input with left icon and right icon 1`] = `
<svg
  aria-hidden="true"
  data-testid="left-icon-test"
  fill="currentColor"
  viewBox="0 0 20 20"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M2.003 5.884L10 9.882l7.997-3.998A2 2 0 0016 4H4a2 2 0 00-1.997 1.884z"
  />
  <path
    d="M18 8.118l-8 4-8-4V14a2 2 0 002 2h12a2 2 0 002-2V8.118z"
  />
</svg>
`;

exports[`TextInput Component should render text input with left icon and right icon 2`] = `
<svg
  aria-hidden="true"
  data-testid="right-icon-test"
  fill="currentColor"
  viewBox="0 0 20 20"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    clip-rule="evenodd"
    d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-7 4a1 1 0 11-2 0 1 1 0 012 0zm-1-9a1 1 0 00-1 1v4a1 1 0 102 0V6a1 1 0 00-1-1z"
    fill-rule="evenodd"
  />
</svg>
`;

exports[`TextInput Component should render text input with success validation 1`] = `
<div
  class="TextInput___StyledDiv7-sc-qqk5wt-11 lhKHte"
  data-testid="success-test-id"
>
  <svg
    aria-hidden="true"
    class="TextInput___StyledCheckCircleIcon-sc-qqk5wt-12 cKHkQF"
    fill="currentColor"
    viewBox="0 0 20 20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clip-rule="evenodd"
      d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.707-9.293a1 1 0 00-1.414-1.414L9 10.586 7.707 9.293a1 1 0 00-1.414 1.414l2 2a1 1 0 001.414 0l4-4z"
      fill-rule="evenodd"
    />
  </svg>
</div>
`;
